restli-common/src/main/java/com/linkedin/restli/internal/common/URIMaskUtil.java

```diff
@@ -121,161 +121,159 @@ public static MaskTree decodeMaskUriFormat(StringBuilder toparse) throws Illegal
     ParseState state = ParseState.PARSE_FIELDS;
 
     DataMap result = new DataMap();
-    Deque<DataMap> stack = new ArrayDeque<DataMap>();
+    Deque<DataMap> stack = new ArrayDeque<>();
     stack.addLast(result);
+    int pos = 0;
 
-    while (toparse.length() > 0)
+    while (pos < toparse.length())
     {
       switch (state)
       {
-      case TRAVERSE:
-        if (toparse.indexOf(",") != 0)
-        {
-          throw new IllegalStateException("Internal Error parsing mask: unexpected parse buffer '"
-              + toparse + "' while traversing");
-        }
-        toparse.delete(0, 1);
-        state = ParseState.PARSE_FIELDS;
-        break;
-      case DESCEND:
-        if (toparse.indexOf(":(") != 0)
-        {
-          throw new IllegalStateException("Internal Error parsing mask: unexpected parse buffer '"
-              + toparse + "' while descending");
-        }
-        toparse.delete(0, 2);
-        state = ParseState.PARSE_FIELDS;
-        break;
-      case PARSE_FIELDS:
+        case TRAVERSE:
+          if (toparse.charAt(pos) != ',')
+          {
+            throw new IllegalStateException("Internal Error parsing mask: unexpected parse buffer '"
+                + toparse + "' while traversing");
+          }
+          pos++;
+          state = ParseState.PARSE_FIELDS;
+          break;
+        case DESCEND:
+          if (toparse.charAt(pos) != ':' || toparse.charAt(pos+1) != '(')
```

Comment:
is there a risk of StringIndexOutOfBounds here when pos + 1 == length

Reply:
Good call